mention/src/parse/error.rs

```diff
@@ -101,101 +90,167 @@ impl Display for ParseMentionError<'_> {
 
 impl Error for ParseMentionError<'_> {
     fn source(&self) -> Option<&(dyn Error + 'static)> {
-        match self {
-            Self::IdNotU64 { source, .. } => Some(source),
-            Self::LeadingArrow { .. }
-            | Self::PartMissing { .. }
-            | Self::Sigil { .. }
-            | Self::TrailingArrow { .. } => None,
-        }
+        self.source
+            .as_ref()
+            .map(|source| &**source as &(dyn Error + 'static))
     }
 }
 
+/// Type of [`ParseMentionError`] that occurred.
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum ParseMentionErrorType<'a> {
+    /// ID portion of the mention isn't a u64.
+    IdNotU64 {
+        /// String that could not be parsed into a u64.
+        found: &'a str,
+    },
+    /// Leading arrow (`<`) is not present.
+    LeadingArrow {
+        /// Character that was instead found where the leading arrow should be.
+        found: Option<char>,
+    },
+    /// One or more parts of the mention are missing.
+    ///
+    /// For example, an emoji mention - `<:name:id>` - has two parts: the `name`
+    /// and the `id`, separated by the sigil (`:`). If the second sigil denoting
+    /// the second part can't be found, then it is missing.
+    PartMissing {
+        /// Number of parts that are expected.
+        expected: usize,
+        /// Number of parts that have been found.
+        found: usize,
+    },
+    /// Mention's sigil is not present.
+    ///
+    /// Users, for example, have the sigil `@`.
+    Sigil {
+        /// Possible sigils that were expected for the mention type.
+        expected: &'a [&'a str],
+        /// Character that was instead found where the sigil should be.
+        found: Option<char>,
+    },
+    /// Trailing arrow (`>`) is not present.
+    TrailingArrow {
+        /// Character that was instead found where the trailing arrow should be.
+        found: Option<char>,
+    },
+}
+
 #[cfg(test)]
 mod tests {
-    use super::ParseMentionError;
+    use super::{ParseMentionError, ParseMentionErrorType};
     use static_assertions::{assert_fields, assert_impl_all};
     use std::{error::Error, fmt::Debug};
 
-    assert_fields!(ParseMentionError::IdNotU64: found, source);
-    assert_fields!(ParseMentionError::LeadingArrow: found);
-    assert_fields!(ParseMentionError::Sigil: expected, found);
-    assert_fields!(ParseMentionError::TrailingArrow: found);
-    assert_impl_all!(ParseMentionError<'_>: Clone, Debug, Error, Eq, PartialEq, Send, Sync);
+    assert_fields!(ParseMentionErrorType::IdNotU64: found);
+    assert_fields!(ParseMentionErrorType::LeadingArrow: found);
+    assert_fields!(ParseMentionErrorType::Sigil: expected, found);
+    assert_fields!(ParseMentionErrorType::TrailingArrow: found);
+    assert_impl_all!(ParseMentionErrorType<'_>: Debug, Send, Sync);
+    assert_impl_all!(ParseMentionError<'_>: Debug, Error, Send, Sync);
```

Comment:
Similarily here, not that `PartialEq` is needed though I want to check if the removal was intentional.

Reply:
There is an inner source error, so guaranteeing that the source error can be equated is probably too much of a commitment.